src/keys/keypair.rs

```diff
@@ -0,0 +1,241 @@
+        let keypair = BlsKeypair {
+            secret: SerdeSecret(bls_secret_key),
+            public: bls_public_key,
+        };
+        Self::Bls(keypair)
+    }
+
+    /// Constructs a BLS public keypair share.
+    pub fn new_bls_share(
+        index: usize,
+        secret_share: threshold_crypto::SecretKeyShare,
+        public_key_set: threshold_crypto::PublicKeySet,
+    ) -> Self {
+        let public_share = secret_share.public_key_share();
+        let keypair_share = BlsKeypairShare {
+            index,
+            secret: SerdeSecret(secret_share),
+            public: public_share,
+            public_key_set,
+        };
+        Self::BlsShare(keypair_share)
+    }
+
+    /// Returns the public key associated with this keypair.
+    pub fn public_key(&self) -> PublicKey {
+        match self {
+            Self::Ed25519(keypair) => PublicKey::Ed25519(keypair.public),
+            Self::Bls(keypair) => PublicKey::Bls(keypair.public),
+            Self::BlsShare(keypair) => PublicKey::BlsShare(keypair.public),
+        }
+    }
+
+    /// Returns the secret key associated with this keypair.
+    pub fn secret_key(&self) -> Result<SecretKey> {
+        match self {
+            Self::Ed25519(keypair) => {
+                let bytes = keypair.secret.to_bytes();
+                match ed25519_dalek::SecretKey::from_bytes(&bytes) {
+                    Ok(sk) => Ok(SecretKey::Ed25519(sk)),
+                    Err(_) => Err(Error::Unexpected(
+                        "Could not deserialise Ed25519 secret key".to_string(),
+                    )),
+                }
+            }
+            Self::Bls(keypair) => Ok(SecretKey::Bls(keypair.secret.clone())),
+            Self::BlsShare(keypair) => Ok(SecretKey::BlsShare(keypair.secret.clone())),
+        }
+    }
+
+    /// Signs with the underlying keypair.
+    pub fn sign(&self, data: &[u8]) -> Signature {
+        match self {
+            Self::Ed25519(keypair) => Signature::Ed25519(keypair.sign(&data)),
+            Self::Bls(keypair) => Signature::Bls(keypair.secret.sign(data)),
+            Self::BlsShare(keypair) => {
+                let index = keypair.index;
+                let share = keypair.secret.sign(data);
+                Signature::BlsShare(SignatureShare { index, share })
+            }
+        }
+    }
+}
+
+impl From<threshold_crypto::SecretKey> for Keypair {
+    fn from(sk: threshold_crypto::SecretKey) -> Self {
+        let public = sk.public_key();
+        let keypair = BlsKeypair {
+            secret: SerdeSecret(sk),
+            public,
+        };
+        Self::Bls(keypair)
+    }
+}
+
+impl From<SerdeSecret<threshold_crypto::SecretKey>> for Keypair {
+    fn from(sk: SerdeSecret<threshold_crypto::SecretKey>) -> Self {
+        let public = sk.public_key();
+        let keypair = BlsKeypair { secret: sk, public };
+        Self::Bls(keypair)
+    }
+}
+
+impl From<ed25519_dalek::SecretKey> for Keypair {
```

Comment:
This one is unusual perhaps as Dalek holds a Keypair type that already has secret and public in it?

Reply:
It does, but it's not our abstracted keypair. Which can be either bls/share or ed